### ISSUE TITLE
Getting `concurrent.futures._base.CancelledError` from simple binary tree built from futures

### ISSUE THREAD

**Issue by cjnolet** [dask/dask#6027]
I'm getting the following exception from a binary tree that I'm building from futures (based largely on an example in the dask docs that uses delayed instead of futures).

Here's the function in question:
```

def tree_reduce(objs, func=sum):

    while len(objs) > 1:
        new_objs = []
        n_objs = len(objs)
        for i in range(0, n_objs, 2):
            inputs = objs[i:i + 2]
            obj = get_client().submit(func, inputs)
            new_objs.append(obj)
        wait(new_objs)
        objs = new_objs

    return first(objs)
```

And my reproducible test:
```
    client = Client(cluster)

    a = client.scatter(range(n_parts))
    b = tree_reduce(a)
    b = b.result()

    assert(sum(range(n_parts)) == b)
```

The exception is intermittent and happens about 50% of the time:
```
self = <tornado.gen.Runner object at 0x7f2cf6ce5710>

    def run(self) -> None:
        """Starts or resumes the generator, running until it reaches a
        yield point that is not ready.
        """
        if self.running or self.finished:
            return
        try:
            self.running = True
            while True:
                future = self.future
                if future is None:
                    raise Exception("No pending future")
                if not future.done():
                    return
                self.future = None
                try:
                    exc_info = None
    
                    try:
>                       value = future.result()
E                       concurrent.futures._base.CancelledError
```

I've tried doing a `dask.distributed.wait` after each level of the tree. I've also tried waiting for `b` in the reproducible example before calling `result`. I'm completely stumped as to why this is happening. I'm running this on a workstation with 2x V100s.

**Comment by mrocklin** [comment]
Thanks for the issue @cjnolet .  I'm sorry to hear that you've had a frustrating time.  

If you can provide a minimal reproducible example I'd be happy to run through it and see if I can reproduce the problem.  You're really close to this with your current code, but it looks like you have a custom function `func` in there, it's not clear how you're creating a cluster (hopefully a bare `LocalCluster()` will do), and there are also missing imports.  Anything you can do to make this process copy-pastable would be welcome. 

You mention that you are using two V100s.  Is it necessary to use GPUs to reproduce the problem?  If so, I might divert you to RAPIDS folks.

**Comment by cjnolet** [comment]
Absolutely, @mrocklin. Actually, I am using the `LocalCUDACluster` but this is a simple pytest that's not storing any data on GPUs (though I was able to reproduce it w/ CuPy arrays). 

Oddly enough, this issue seems exclusive to the use of `client.scatter`. This example seems to work fine: 

```
    from dask_cuda import LocalCUDACluster
    from dask.distributed import Client

    cluster = LocalCUDACluster(protocol="tcp")
    client = Client(cluster)

    import numpy as np

    def s(x):
        return x

    n_parts = 15
    a = [client.submit(s, i) for i in range(n_parts)]
    b = tree_reduce(a)
    b = b.result()

    print(str(b))

    assert(sum(range(n_parts)) == b)
```

**Comment by mrocklin** [comment]
Thanks @cjnolet .  In the future if you're able to provide a copy-pasteable example that anyone can run that would help.  For example, if you don't need `LocalCUDACluster` then please just don't include it in the example.  That way people who don't have access to GPUs (like me) can help you without having to manually modify your example.  Also, if you're able to provide something that a maintainer can just copy and paste into an IPython session it really helps out a lot.  Most of us aren't actually paid to help you with these things, and we go through a lot of issues, so anything you can do to get the time expectation on our end to under a minute is very welcome.

I tried cobbling together a copy-pastable example from your comments below:

```python
from dask.distributed import get_client, wait, Client, LocalCluster
from toolz import first

def tree_reduce(objs, func=sum):

    while len(objs) > 1:
        new_objs = []
        n_objs = len(objs)
        for i in range(0, n_objs, 2):
            inputs = objs[i:i + 2]
            obj = get_client().submit(func, inputs)
            new_objs.append(obj)
        wait(new_objs)
        objs = new_objs

    return first(objs)

cluster = LocalCluster()
client = Client(cluster)

def add(parts): return sum(parts)

n_parts = 15
a = client.scatter(range(n_parts))
b = tree_reduce(a, func=add)
b = b.result()

print(str(b))

assert(sum(range(n_parts)) == b)
```

Unfortunately it seems to work fine on my machine.  I ran this ten times and each time it was fine.  

So I think that the next step here is for you to try to provide a copy-pasteable example that a maintainer can run that fails.  

(Sorry for the github issue lecture.  It's been a busy day.)

**Comment by cjnolet** [comment]
My apologies, @mrocklin. Thank you for being patient with me. 

Here's a slightly modified script that fails for me everytime. It seems the number of workers is important (On my tests, I have only 2 workers and in our environment it's not always trivial to increase the number of workers). 

```python
from dask.distributed import get_client, wait, Client, LocalCluster

from toolz import first

#import pytest

def tree_reduce(objs, func=sum):

    while len(objs) > 1:
        new_objs = []
        n_objs = len(objs)
        for i in range(0, n_objs, 2):
            inputs = objs[i:i + 2]
            obj = get_client().submit(func, inputs)
            new_objs.append(obj)
        wait(new_objs)
        objs = new_objs

    return first(objs)


#@pytest.mark.parametrize("n_parts", [1, 2, 5, 10, 15])
def test(n_parts):

    client = Client(cluster)

    def add(parts): return sum(parts)

    a = client.scatter(range(n_parts))
    b = tree_reduce(a, func=add)
    b = b.result()

    print(str(b))
    assert(sum(range(n_parts)) == b)


if __name__ == "__main__":

    cluster = LocalCluster(n_workers=2)

    for i in range(25):

        for n_parts in [1, 2, 5, 10, 15]:
            test(n_parts)
```

**Comment by mrocklin** [comment]
Ah, that's perfect.  Thank you @cjnolet .  I'm able to reproduce.  I'll try to take a look at this this weekend.

**Comment by mrocklin** [comment]
Here is a simpler failure.  I've removed the tree reduction.

```python
from dask.distributed import Client, LocalCluster


async def test(n_parts, client):
    a = await client.scatter(range(n_parts), hash=True)
    future = client.submit(sum, a)
    await future


async def f():
    async with Client(n_workers=2, processes=False, asynchronous=True) as client:
        for i in range(25):
            for n_parts in [1, 2, 5, 10, 15]:
                print(n_parts)
                await test(n_parts, client)

if __name__ == "__main__":
    import asyncio
    asyncio.get_event_loop().run_until_complete(f())
```

I think that the problem here has to do with the scattered futures, which get deleted and recreated in quick succession.  There is probably some subtle race condition in how the client and scheduler are counting references.  @jacobtomlinson , this might be something that you enjoy digging into.  It should give some understanding of internals, and should help out other NVIDIA folks.

**Comment by cjnolet** [comment]
Thanks @mrocklin! I verified that your new script fails every time when `n_parts=2`

Our CI always executes our tests against the bleeding edge versions of dask and distributed. It is currently blocked by the sudden presence of `CancelledError` in several of our algorithms. It seems that these errors started about a week ago and we're seeing the issue even in tests / code that haven't change recently.

In addition to the `cancelledError`, I'm also seeing some strange subsequent errors in dask array/cudf `from_delayed()` that an instance of a `cudf.Dataframe` was expected but a `Future` was found instead. Sometimes I'm able to fix these errors by not having a `future.result()` overwrite a variable that was holding a reference to a future, for example:

This consistently has a cancelled error:
```
X = [client.submit(...) for x in X]
X = client.compute(X, sync=True)
```

This errors less frequently:
```
X = [client.submit(...) for x in X]
X_local = client.compute(X, sync=True)
```

Based on the example, I'm thinking this workaround is not fixing the issue, but just lowering the chances for the potential race condition.

**Comment by mrocklin** [comment]
The error that you have uncovered here happens when 

1.  We create some futures by scattering data (not normal computations)
2.  We delete those futures
3.  We recreate those exact same futures a millisecond later

I don't have enough information on your other fail cases to know if they are the same unfortunately.  If you're not scattering things then it might be something different.  In that case I would encourage you to raise another issue with a minimal reproducible example.

**Comment by quasiben** [comment]
I looked at this and there is a race condition but I am not sure that we can easily resolve inside of dask.  The race condition being that as we are scattering, we are deleting at the same.  I think there are a few options:

1) when we scatter wait a tick
2) when we scatter call transitions directly (it's not this call but something like, scheduler -- complete your tasks
3) ask user to wait a tick in between scatter/submit calls
4) uniqueness around scattered keys (this could be fairly limiting for other tasks but would avoid collisions)

**Comment by mrocklin** [comment]
Dask should be robust to this.  We shouldn't rely on users to be careful
here.

I think that someone needs to look deeply at how we currently count
references and send task-released messages around.  I think that
@jacobtomlinson would be a good person for this if he has time.

**Comment by mrocklin** [comment]
In other words.  This isn't something that we want to work around or
avoid.  This is something that we should dive into.  Diving into it is hard
though and will require some previous experience with the internals of the
distributed scheduler.

**Comment by jacobtomlinson** [comment]
I am looking at this now. Although I'm feeling a little in at the deep end with this. Some pointers to where I should start looking would be appreciated.

**Comment by mrocklin** [comment]
Yeah, it's non-trivial.  I would probably start by printing out or
otherwise recording every time the client and scheduler think that the task
has been created or cancelled.  I think that by looking through that list
that you'll be able to eventually spot something fishy.

**Comment by jacobtomlinson** [comment]
I've created another further simplified reproducer.

```python
from dask.distributed import Client


async def f():
    async with Client(
        n_workers=1, processes=False, asynchronous=True, silence_logs=False
    ) as client:

        d = [0, 1]
        a = await client.scatter(d, hash=True)
        print(await client.submit(sum, a))

        del a

        d = [1, 2]
        a = await client.scatter(d, hash=True)
        print(await client.submit(sum, a))


if __name__ == "__main__":
    import asyncio

    asyncio.get_event_loop().run_until_complete(f())
```

If we scatter some data and delete it, but then immediately scatter some more data with a common value and try to operate on it we see the `CancelledError`.

This is because the following chain of events happen:

- Futures are created by the client and data is sent to the scheduler
- Scheduler creates the keys and sends the data to the worker
- Futures are deleted by the client, queueing a request to the scheduler for deletion
- Scheduler queues keys for deletion by the workers
- New futures are created by the client
- Scheduler creates new keys, one is the same as the previously deleted key, and sends data to the worker
- Deletion task arrives at the worker and it removes the keys, despite one being recreated
- Scheduler tries to compute on lost data, the exception is raised

The issue seems to be a result of the time delay between the client deleting a future and the worker deleting the data. The data is recreated during that time.

I feel like the logic here is reasonable. If the user deletes some data it should be deleted. If they recreate it again then they pay a performance penalty compared with not deleting it.

Perhaps one solution to this would be to introduce some reference counting for keys on the worker. When a key is created in `Worker.update_data` a count should be incremented. And then when it is deleted in `Worker.delete_data` it should be decremented and deleted only if the count gets to zero.

I'll start working on a PR for this.

**Comment by kumarprabhu1988** [comment]
Is this issue also present in dask-yarn cluster? I tried out the easy reproducible example and see the same error. If yes, would this be fixed by https://github.com/dask/distributed/pull/3643?

**Comment by dkhokhlov** [comment]
Getting the same  concurrent.futures._base.CancelledError error on LocalCluster.
I am using client.run() to run a function on all workers in sync mode. Error is intermittent, 1 in 10 runs.
I am wondering if this is related to nanny. By default run does not use it.
